File: src/pvz_console/cli/menus.py
from __future__ import annotations

import os
from dataclasses import dataclass
from typing import List, Optional, Sequence, Union

from pvz_console.cli.theme import THEME
from pvz_console.parsers.loaders import list_localizations

HEADER_FILL = "\u2500"
ALL_TOKEN = "*"
PROMPT_SYMBOL = "\u276f"


@dataclass(frozen=True)
class MenuOption:
    key: str
    label: str


def clear_console() -> None:
    os.system("cls" if os.name == "nt" else "clear")


def press_enter_to_continue() -> None:
    input("\n  Press Enter to continue... ")
    clear_console()


def _parse_int(raw: str, default: int) -> int:
    try:
        return int(raw.strip())
    except ValueError:
        return default


def _blank_lines(n: int = 1) -> None:
    for _ in range(THEME.blank_lines * n):
        print()


def prompt() -> str:
    return input(f"  {THEME.accented(PROMPT_SYMBOL, bold=True)} ")


def section(title: str) -> None:
    _blank_lines()
    print(f"  {THEME.accented(title.upper(), bold=True)}")
    print(f"  {THEME.accented(HEADER_FILL * len(title))}")
    _blank_lines()


def print_menu(title: str, options: Sequence[MenuOption]) -> None:
    section(title)
    key_width = max(len(opt.key) for opt in options)
    for opt in options:
        key_fmt = f"{opt.key:>{key_width}}"
        print(f"    [{THEME.accented(key_fmt, bold=True)}]  {THEME.primary(opt.label)}")
    _blank_lines()


def ask_choice(title: str, options: Sequence[MenuOption], default: int = -1) -> int:
    print_menu(title, options)
    return _parse_int(prompt(), default)


def info(message: str) -> None:
    print(f"  {THEME.primary(message)}")


def warn(message: str) -> None:
    icon = THEME.emoji("\u26a0\ufe0f ", "[!]")
    print(f"  {icon} {THEME.colorize(message, 'yellow')}")


def success(message: str) -> None:
    icon = THEME.emoji("\u2705", "[OK]")
    print(f"  {icon} {THEME.colorize(message, 'green')}")


def error(message: str) -> None:
    icon = THEME.emoji("\u274c", "[X]")
    print(f"  {icon} {THEME.colorize(message, 'red')}")


# ---------- localization picker -------------------------------------------------


def _render_locales_grid(locales: Sequence[str], columns: int = 3) -> None:
    width = max((len(loc) for loc in locales), default=0)
    rows = (len(locales) + columns - 1) // columns
    for row in range(rows):
        cells = []
        for col in range(columns):
            idx = row + col * rows
            if idx >= len(locales):
                continue
            key = THEME.accented(f"{idx:>2}", bold=True)
            cells.append(f"[{key}]  {locales[idx]:<{width}}")
        print("    " + "   ".join(cells))


def select_localization(
    root: str,
    *,
    allow_multi: bool = True,
    exclude: Optional[List[str]] = None,
) -> Union[str, List[str]]:
    """Prompt for a locale. Returns a single locale name, or the full list when
    ``allow_multi`` is true and the user picks ``*``.
    """
    section("Select a localization")
    localizations = list_localizations(root)
    if exclude:
        localizations = [loc for loc in localizations if loc not in exclude]

    if allow_multi:
        print(f"    [ {THEME.accented(ALL_TOKEN, bold=True)}]  All locales")
        _blank_lines()
    _render_locales_grid(localizations)
    hint = f"(number or '{ALL_TOKEN}' for all)" if allow_multi else "(single locale only)"
    _blank_lines()
    info(hint)

    while True:
        raw = prompt().strip()
        if allow_multi and raw == ALL_TOKEN:
            return list(localizations)
        idx = _parse_int(raw, -1)
        if 0 <= idx < len(localizations):
            return localizations[idx]
        error("Invalid choice. Please enter a number from the list"
              + (f" or '{ALL_TOKEN}'." if allow_multi else "."))


def ask_text(label: str, default: str = "") -> str:
    suffix = f" [{default}]" if default else ""
    raw = input(f"  {THEME.primary(label + suffix + ': ')}")
    return raw.strip() or default


def ask_choice_from_list(label: str, values: Sequence[str], current: str) -> str:
    """Show a short list; return the chosen value or the current one on invalid input."""
    section(label)
    for idx, value in enumerate(values):
        marker = " (current)" if value == current else ""
        key = THEME.accented(f"{idx:>2}", bold=True)
        print(f"    [{key}]  {value}{marker}")
    _blank_lines()
    idx = _parse_int(prompt(), -1)
    if 0 <= idx < len(values):
        return values[idx]
    return current

File: src/cli/menus.ts
/** The interactive prompts, built on `@clack/prompts`.
 *
 * ## Testability seam
 *
 * Clack puts the terminal in raw mode, so the screen-level tests never reach
 * this module: they inject an `AppDeps` bundle of stubs instead (see
 * `cli/deps.ts`). What is exercised here directly is the translation layer we
 * own — cancellation mapping, the "all locales" sentinel, the picker's
 * thresholds — by handing {@link ConsoleIO} a pair of in-memory streams.
 *
 * ## Cancellation
 *
 * `Esc` / `Ctrl+C` resolve clack prompts with a cancel symbol rather than
 * killing the process. Each wrapper maps that to the least destructive
 * outcome — {@link MENU_CANCELLED}, the current value, `null` for the locale
 * picker — so every screen is escapable and the main menu's *Exit* is the only
 * way out. Nothing is ever written on a cancelled prompt.
 */

import { autocomplete, confirm, isCancel, select, text } from "@clack/prompts";

import { type ConsoleIO, defaultIO, error, streams } from "./output";
import { THEME } from "./theme";
import { listLocalizations } from "../parsers/loaders";

/** Sentinel value of the "All locales" entry in the localization picker. */
export const ALL_TOKEN = "*";
/**
 * Returned by {@link askChoice} when the user escapes the menu.
 *
 * A dedicated value rather than the menu's *Back* key: a menu is free to give
 * `0` a meaning of its own (`Show what's missing` uses it for *All types*), and
 * escaping must never trigger an action.
 */
export const MENU_CANCELLED = -1;
/** Above this many locales the picker becomes a type-to-filter autocomplete. */
const AUTOCOMPLETE_THRESHOLD = 10;
/** Rows shown before a list starts scrolling. */
const MAX_VISIBLE_ITEMS = 12;

export interface MenuOption {
  key: string;
  label: string;
  /** Short annotation shown next to the highlighted row. */
  hint?: string;
}

function optionValue(option: MenuOption, index: number): number {
  const parsed = Number.parseInt(option.key, 10);
  return Number.isNaN(parsed) ? index : parsed;
}

/** Where a menu starts, and which streams it talks to. */
export interface ChoicePrompt {
  initialValue?: number;
  io?: ConsoleIO;
}

/** Pick a menu entry; returns its numeric key, or {@link MENU_CANCELLED}. */
export async function askChoice(
  title: string,
  options: readonly MenuOption[],
  prompt: ChoicePrompt = {},
): Promise<number> {
  const io = prompt.io ?? defaultIO;
  const choice = await select<number>({
    message: THEME.heading(title),
    options: options.map((option, index) => ({
      value: optionValue(option, index),
      label: THEME.primary(option.label),
      hint: option.hint,
    })),
    initialValue: prompt.initialValue,
    maxItems: MAX_VISIBLE_ITEMS,
    ...streams(io),
  });
  return isCancel(choice) ? MENU_CANCELLED : choice;
}

/** The value a list prompt starts on, and which streams it talks to. */
export interface ListPrompt {
  current: string;
  io?: ConsoleIO;
}

/** Pick from a short list of values; cancelling keeps `current`. */
export async function askChoiceFromList(
  label: string,
  values: readonly string[],
  prompt: ListPrompt,
): Promise<string> {
  const io = prompt.io ?? defaultIO;
  const choice = await select<string>({
    message: THEME.heading(label),
    options: values.map((value) => ({
      value,
      label: value,
      hint: value === prompt.current ? "current" : undefined,
    })),
    initialValue: prompt.current,
    maxItems: MAX_VISIBLE_ITEMS,
    ...streams(io),
  });
  return isCancel(choice) ? prompt.current : choice;
}

export async function askText(
  label: string,
  defaultValue = "",
  io: ConsoleIO = defaultIO,
): Promise<string> {
  const answer = await text({
    message: THEME.heading(label),
    placeholder: defaultValue || undefined,
    defaultValue,
    ...streams(io),
  });
  if (isCancel(answer)) {
    return defaultValue;
  }
  return answer.trim() || defaultValue;
}

export async function askConfirm(
  label: string,
  defaultValue = false,
  io: ConsoleIO = defaultIO,
): Promise<boolean> {
  const answer = await confirm({
    message: THEME.heading(label),
    initialValue: defaultValue,
    ...streams(io),
  });
  return isCancel(answer) ? defaultValue : answer;
}

// ---------- localization picker -------------------------------------------------

export interface SelectLocalizationOptions {
  allowMulti?: boolean;
  exclude?: readonly string[] | null;
}

interface PickerOption {
  value: string;
  label: string;
  hint?: string;
}

function pickerOptions(locales: readonly string[], allowMulti: boolean): PickerOption[] {
  const all = { value: ALL_TOKEN, label: "All locales", hint: `${locales.length} folders` };
  return [
    ...(allowMulti ? [all] : []),
    ...locales.map((locale) => ({ value: locale, label: locale })),
  ];
}

/** Long lists get a type-to-filter prompt: two dozen locales are slow to arrow through. */
function promptForLocale(
  options: PickerOption[],
  localeCount: number,
  io: ConsoleIO,
): Promise<string | symbol> {
  const message = THEME.heading("Select a localization");
  const shared = { message, options, maxItems: MAX_VISIBLE_ITEMS, ...streams(io) };
  return localeCount > AUTOCOMPLETE_THRESHOLD
    ? autocomplete<string>({ ...shared, placeholder: "type to filter…" })
    : select<string>(shared);
}

/**
 * Prompt for a locale: a single name, the full list when the user picks
 * *All locales*, or `null` when the picker is cancelled or has nothing to show.
 */
export async function selectLocalization(
  root: string,
  { allowMulti = true, exclude = null }: SelectLocalizationOptions = {},
  io: ConsoleIO = defaultIO,
): Promise<string | string[] | null> {
  const excluded = new Set(exclude ?? []);
  const locales = listLocalizations(root).filter((locale) => !excluded.has(locale));

  if (locales.length === 0) {
    error("No localization folder found under the configured project root.", io);
    return null;
  }

  const choice = await promptForLocale(
    pickerOptions(locales, allowMulti),
    locales.length,
    io,
  );

  if (isCancel(choice)) {
    return null;
  }
  return choice === ALL_TOKEN ? [...locales] : choice;
}
